contracts/fionft.sol

```diff
@@ -117,193 +94,196 @@ contract FIONFT is ERC721, ERC721Pausable {
-      require(bytes(obtid).length > 0, "Invalid obtid");
-      bytes32 obthash = keccak256(bytes(abi.encode(obtid)));
-      return (approvals[obthash].approvals, approvals[obthash].account, approvals[obthash].obtid);
+    function getOracle(address account) external view returns (bool, uint32) {
+      require(account != address(0), "Invalid account");
+      return (hasRole(ORACLE_ROLE, account), uint32(oraclelist.length));
+    }
+
+    function getApproval(bytes memory obtid) external view returns (uint32, bool) {
+      require(obtid.length > 0, "Invalid obtid");
+      return (approvals[bytes32(obtid)].approvals, approvals[bytes32(obtid)].complete);
+    }
+
+    function listDomainsOfOwner(address _owner) public view returns(string[] memory ownerTokens) {
+    uint256 tokenCount = balanceOf(_owner);
+    if (tokenCount == 0) {
+        return new string[](0);
+    } else {
+        string[] memory result = new string[](tokenCount);
+        uint256 resultIndex = 0;
+        uint256 tokenId;
+        for (tokenId = 1; tokenId <= _tokenIds._value; tokenId++) {
+            if (_owners[tokenId] == _owner) {
+                result[resultIndex] = attribute[tokenId];
+                resultIndex++;
+            }
+        }
+            return result;
+        }
     }
 
     function getOracles() external view returns(address[] memory) {
       return oraclelist;
     }
 
-    function regoracle(address account) external custodianOnly {
-      require(account != address(0), "Invalid address");
+    function regoracle(address account) external onlyRole(CUSTODIAN_ROLE)  {
+      require(account != address(0), "Invalid account");
       require(account != msg.sender, "Cannot register self");
-      require(!oracles[account].active, "Oracle already registered");
-      bytes32 id = keccak256(bytes(abi.encode("ro",account, roracmapv )));
-      require(!approvals[id].approved[msg.sender],  "Already approved");
-      int reqcust = custodian_count * 2 / 3 + 1;
-      if (approvals[id].approvals < reqcust) {
-        approvals[id].approvals++;
-        approvals[id].approved[msg.sender] = true;
-      }
-      if (approvals[id].approvals == reqcust){
-        oracles[account].active=true;
+      require(!hasRole(ORACLE_ROLE, account), "Already registered");
+      bytes32 indexhash = keccak256(bytes(abi.encode(ApprovalType.AddOracle,account )));
+      if (getConsensus(indexhash, 1)){
+        _grantRole(ORACLE_ROLE, account);
         oracle_count++;
         oraclelist.push(account);
-        delete approvals[id];
-        roracmapv++;
-        emit oracle_registered(account, id);
+        emit oracle_registered(account, indexhash);
       }
+      emit consensus_activity("custodian", indexhash);
     }
 
-    function unregoracle(address account) external custodianOnly {
-      require(account != address(0), "Invalid address");
+    function unregoracle(address account) external onlyRole(CUSTODIAN_ROLE) {
```

Comment:
This needs to support: "A minimum of 3 oracles must be maintained. If only 3 Oracles are registered, unregoracle can not be called."
So, unregoracle should only be callable if there are 4 or more oracles registered.

Reply:
Fixed.